src/Utils/QueryAnalyzer.php

```diff
@@ -0,0 +1,489 @@
+
+		// If query analyzer is disabled, bail
+		if ( true !== $should_analyze_queries ) {
+			return;
+		}
+
+		// track keys related to the query
+		add_action( 'do_graphql_request', [ $this, 'determine_graphql_keys' ], 10, 4 );
+
+		// Track models loaded during execution
+		add_filter( 'graphql_dataloader_get_model', [ $this, 'track_nodes' ], 10, 1 );
+
+	}
+
+	/**
+	 * Determine the keys associated with the GraphQL document being executed
+	 *
+	 * @param ?string         $query     The GraphQL query
+	 * @param ?string         $operation The name of the operation
+	 * @param ?array          $variables Variables to be passed to your GraphQL request
+	 * @param OperationParams $params    The Operation Params. This includes any extra params, such as extenions or any other modifications to the request body
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function determine_graphql_keys( ?string $query, ?string $operation, ?array $variables, OperationParams $params ): void {
+
+		// @todo: support for QueryID?
+
+		// if the query is empty, get it from the request params
+		if ( empty( $query ) ) {
+			$query = $this->request->params->query ?: null;
+		}
+
+		if ( empty( $query ) ) {
+			return;
+		}
+
+		$query_id       = Utils::get_query_id( $query );
+		$this->query_id = $query_id ?: uniqid( 'gql:', true );
+
+		// if there's a query (either saved or part of the request params)
+		// get the GraphQL Types being asked for by the query
+		$this->list_types    = $this->set_list_types( $this->schema, $query );
+		$this->queried_types = $this->set_query_types( $this->schema, $query );
+		$this->models        = $this->set_query_models( $this->schema, $query );
+
+		/**
+		 * @param QueryAnalyzer $query_analyzer The instance of the query analyzer
+		 * @param string $query The query string being executed
+		 */
+		do_action( 'graphql_determine_graphql_keys', $this, $query );
+
+	}
+
+	/**
+	 * @return array
+	 */
+	public function get_list_types(): array {
+		return $this->list_types;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function get_query_types(): array {
+		return $this->queried_types;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function get_query_models(): array {
+		return $this->models;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function get_runtime_nodes(): array {
+		return $this->runtime_nodes;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function get_root_operation(): string {
+		return $this->root_operation;
+	}
+
+	/**
+	 * @return string|null
+	 */
+	public function get_query_id(): ?string {
+		return $this->query_id;
+	}
+
+	/**
+	 * Given the Schema and a query string, return a list of GraphQL Types that are being asked for
+	 * by the query.
+	 *
+	 * @param ?Schema $schema The WPGraphQL Schema
+	 * @param ?string $query  The query string
+	 *
+	 * @return array
+	 * @throws SyntaxError|Exception
+	 */
+	public function set_list_types( ?Schema $schema, ?string $query ): array {
+
+		/**
```

Comment:
Not sure if the docblocks are used for auto-generation, but if so I think this is a bit more intuitive:
```php
/**
 * @param array|null $list_types. Defaults to null.
 * @param ?Schema $schema The WPGraphQL Schema for the current request
 * @param ?string $query The query string being requested
 */
 $pre_get_list_types = apply_filters( 'graphql_pre_query_analyzer_get_list_types, null, $shema, $query );
 ```
 
 Same for the the other filters using `$null` in this file.
 
 Eitther way the first param should be `array|null` for advanced static analysis.